### ISSUE TITLE
ThreadError: deadlock when lookbook is opened

### ISSUE THREAD

**Issue by vojtad** [ViewComponent/lookbook#218]
### Describe the bug

When loading lookbook the Rails app throws `ThreadError` (part of the log pasted below). Every subsequent request after the error is reported for the first time fails with the same error.

It seems the error is thrown right after lookbook opens a WebSocket connection with Rails server.

```
21:26:00 server.1 | [43892492-2826-4fc3-ab79-a58838b41c19] Started GET "/lookbook-assets/feather-sprite.svg" for 127.0.0.1 at 2022-10-31 20:26:00 +0000
21:26:00 server.1 | [58ef297c-8a48-4172-8fb1-2f3829ca8726] Started GET "/lookbook/cable?uid=1667247959878" for 127.0.0.1 at 2022-10-31 20:26:00 +0000
21:26:00 server.1 | [58ef297c-8a48-4172-8fb1-2f3829ca8726] [user:136] Started GET "/lookbook/cable/?uid=1667247959878" [WebSocket] for 127.0.0.1 at 2022-10-31 20:26:00 +0000
21:26:00 server.1 | [58ef297c-8a48-4172-8fb1-2f3829ca8726] [user:136] Successfully upgraded to WebSocket (REQUEST_METHOD: GET, HTTP_CONNECTION: Upgrade, HTTP_UPGRADE: websocket)
21:26:06 server.1 | 2022-10-31 20:26:06 +0000 Rack app ("GET /lookbook/cableuid=1667247959878" - (127.0.0.1)): #<ThreadError: deadlock; recursive locking>
21:26:14 server.1 | 2022-10-31 20:26:14 +0000 Rack app ("GET /lookbook/cableuid=1667247959878" - (127.0.0.1)): #<ThreadError: deadlock; recursive locking>
```

### To Reproduce

1. We open lookbook main page in our app
2. `ThreadError` appears in Rails log after the first WebSocket request

### Expected behavior

No `ThreadError`, we expect lookbook page to load without any issues.

### Version numbers 

Please complete the following information:

* Lookbook: 1.2.1
* ViewComponent: 2.74.1
* Rails: 7-0-stable branch from a week ago (7.0.4)
* Ruby: 3.1.2
* Puma: 6.0.0

### Additional context

When I increase `PUMA_THREADS` to more than 1 no `ThreadError` is thrown but Rails app gets stuck. I believe deadlock still happens but it is not detected and reported in the log.

I tried enabling `ActionDispatch::DebugLocks ` middleware but `/rails/locks` response is empty.

Here is an output rendered in browser when request is sent to the server when it is deadlocked.

```
Puma caught this error: deadlock; recursive locking (ThreadError)
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rack-2.2.4/lib/rack/lock.rb:14:in `lock'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rack-2.2.4/lib/rack/lock.rb:14:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rack-2.2.4/lib/rack/deflater.rb:44:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/rails-76b16d338b47/actionpack/lib/action_dispatch/middleware/static.rb:23:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rack-2.2.4/lib/rack/sendfile.rb:110:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/rails-76b16d338b47/actionpack/lib/action_dispatch/middleware/ssl.rb:77:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/rails-76b16d338b47/actionpack/lib/action_dispatch/middleware/host_authorization.rb:131:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rack-cors-1.1.1/lib/rack/cors.rb:100:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/shakapacker-6.5.2/lib/webpacker/dev_server_proxy.rb:25:in `perform_request'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rack-proxy-0.7.4/lib/rack/proxy.rb:78:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/scout_apm-5.3.2/lib/scout_apm/instruments/middleware_summary.rb:58:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/rails-76b16d338b47/railties/lib/rails/engine.rb:530:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/configuration.rb:268:in `call'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/request.rb:89:in `block in handle_request'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/thread_pool.rb:343:in `with_force_shutdown'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/request.rb:88:in `handle_request'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/server.rb:430:in `process_client'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/server.rb:233:in `block in run'
/Users/vojtadrbohlav/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/thread_pool.rb:150:in `block in spawn_thread'
```

**Comment by allmarkedup** [comment]
Hey @vojtad thanks for the bug report and the detailed info.

This is not an error I've seen before and unfortunately so far I've been unable to recreate with a fresh install of Rails with those same gem versions. Could you try a couple of things for me to help see if we can figure out what is going on here?

Firstly - can you try disabling the 'auto refresh' feature by setting `config.lookbook.auto_refresh = false` in your app config? This should disable the websocket. Making a change to a component preview or component file and then manually refreshing the Lookbook preview should still show the change but it won't happen automatically after the file is edited. Are you still seeing any ThreadError messages in the logs after that (I assume not?)

Secondly - I notice you are using Shakapacker. Would you be able to disable that temporarily and see if the problem still occurs? I just want to try to rule out any weird interactions with the dev server if possible.

Sorry you are running into problems but hopefully we can get to the bottom of this quickly.

**Comment by vojtad** [comment]
Thanks for the swift response!

Setting `config.lookbook.auto_refresh = false` helped and we no longer see ThreadError messages. I suspected it will be something related to auto reloading and websocket and looked for such setting but I missed it.

This is good-enough-workaround for us right now. Our primary use case for Lookbook is using it as a UI/UX library for our designers. Not for developing new components (at least not right now).

Disabling Shakapacker didn't change anything. I tried disabling other gems like rack-cors, scout_apm but nothing changed either.

I am trying to setup LLDB to see where Puma gets stuck but no luck for now. I haven't been able to get Ruby backtrace from LLDB, yet.

**Comment by vojtad** [comment]
I've found the issue.

We did have `config.allow_concurrency = false` set in development environment. It's there for a very long time and I actually don't know why. I removed it and everything is working perfectly fine now.

`Rack::Lock` middleware gets inserted when `config.allow_concurrency = false` is set. This middleware locks a mutex around each request. My  guess is this breaks when websocket connection is made. Perhaps something to do with `rack.hijack` and mutex not being unlocked because the websocket request didn't actually finish but it was `rack.hijack`ed and is being handled by ActionCable.
